### ISSUE TITLE
Support repeats in relational flat views

### ISSUE THREAD

**Issue by avernet** [orbeon/orbeon-forms#1069]
Right now, this flat view feature is only available for Oracle, and it doesn't support repeats. The current Oracle view is produced from a query of the form:

``` sql
select
    document_id metadata_document_id, created metadata_created, last_modified metadata_last_modified, username metadata_username
        , extractValue(xml, '/*/personal/first-name')
        "PERSONAL_FIRST_NAME"
        , extractValue(xml, '/*/personal/last-name')
        "PERSONAL_LAST_NAME"
        , extractValue(xml, '/*/personal/phone')
        "PERSONAL_PHONE"
        , extractValue(xml, '/*/personal/number')
        "PERSONAL_NUMBER"
        , extractValue(xml, '/*/personal/type')
        "PERSONAL_TYPE"
from (
    select d.*, dense_rank() over (partition by document_id order by last_modified desc) as latest
    from orbeon_form_data d
    where
        app = 'datadec'
        and form = 'repeat-true'
    )
where latest = 1 and deleted = 'N'
```

If we want the same technique to work on all relational databases, and in particular MySQL which doesn't know how to index XML data or support function indices, we most likely have to create other tables where the data

**Comment by evlist** [comment]
+1 from customer (for Postgresql).
  Any update on this RFE ?

**Comment by avernet** [comment]
@evlist It's still on our radar, but there is nothing new on this at this point.

**Comment by acspike** [comment]
I make views similar to the following for each repeat by hand:
```
CREATE OR REPLACE VIEW orbeon_default."AppNameFormNameGridName" AS 
 WITH a AS (
         SELECT d.document_id AS metadata_document_id,
            d.created AS metadata_created,
            d.last_modified_time AS metadata_last_modified_time,
            d.last_modified_by AS metadata_last_modified_by,
            generate_subscripts(xpath('/*/section//grid'::text, d.xml), 1) AS idx,
            unnest(xpath('/*/section//grid'::text, d.xml)) AS elt
           FROM orbeon_default.orbeon_form_data d,
            ( SELECT max(d_1.last_modified_time) AS last_modified_time,
                    d_1.app,
                    d_1.form,
                    d_1.document_id
                   FROM orbeon_default.orbeon_form_data d_1
                  WHERE d_1.app::text = 'AppName'::text AND d_1.form::text = 'FormName'::text AND d_1.draft = 'N'::bpchar
                  GROUP BY d_1.app, d_1.form, d_1.document_id) m
          WHERE d.last_modified_time = m.last_modified_time AND d.app::text = m.app::text AND d.form::text = m.form::text AND d.document_id::text = m.document_id::text AND d.deleted = 'N'::bpchar
        )
 SELECT a.metadata_document_id,
    a.metadata_created,
    a.metadata_last_modified_time,
    a.metadata_last_modified_by,
    a.idx AS metadata_idx,
    (xpath('//control1/text()'::text, a.elt))[1]::character varying AS control1,
    (xpath('//control2/text()'::text, a.elt))[1]::character varying AS control2
   FROM a;
```

**Comment by evlist** [comment]
@acspike Thanks for the example !

**Comment by hoila** [comment]
Last year a trainee made a java program to create the views for the repeated sections and the repeated grids for oracle.
 He used the xmltable function.
 For the primary key we use the id field of the orbeon_form_data and the rownum.
 The subsections field contains the xml of the repeated grid, in case the repeated section contains also a repeated grid.
 
 The orbeon form builder doesn't permit duplicate control  names. So we use in the form  for the control names the field names of the table and use then in the view the control names.
 
Example sql for a repeated section

>  CREATE OR REPLACE VIEW test_view_repeated_section  AS 
  SELECT  rownum AS SECTION_NUMBER
        , x.id AS orbeon_form_data
		, xt."FIELD1"
		, xt."FIELD2"
		, xt."SUBSECTIONS"  
	FROM orbeon.orbeon_form_data x,   
	  	 XMLTABLE('/form/section-1/section-1-iteration' PASSING x.xml 
		 COLUMNS           
		    FIELD1    VARCHAR2(20)  PATH 'field1'
		  , FIELD2    VARCHAR2(20)  PATH 'field2'
		  , SUBSECTIONS XMLType PATH '*[not(count(*)=0)]'         ) xt
		  , ( SELECT max(last_modified_time) last_modified_time
		           , app
			   , form
			   , document_id              
			  FROM   orbeon.orbeon_form_data d             
			  WHERE  app   = 'appname'                     
			     AND form  = 'formname' 
			     AND draft = 'N' 
 			  GROUP BY app, form, document_id  ) m 
	WHERE   x.last_modified_time = m.last_modified_time        
	    AND x.app                = m.app        
	    AND x.form               = m.form        
	    AND x.document_id   = m.document_id        
	    AND x.deleted            = 'N';

**Comment by ebruchez** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#drafts/FMfcgxvxBXtHXwGMmHHqvgNnpclRDXfg)

**Comment by ebruchez** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#inbox/FMfcgxwCgVSRczlgzVzTmjnKBhBWcjnK)

**Comment by ebruchez** [comment]
The Multiple File Attachments control is not supported either.

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/18358270/messages/6079549706)

**Comment by obruchez** [comment]
- [ ] doc

**Comment by obruchez** [comment]
- [ ] understand why we have a problem with PostgreSQL 9.5 (which is unsupported since 2021)

**Comment by avernet** [comment]
Form to test names generated for the views and their columns: 

To get the list of the DROP statements, run:

```sql
SELECT 'DROP VIEW ' + QUOTENAME(SCHEMA_NAME(schema_id)) + '.' + QUOTENAME(name) + ';'
FROM sys.views 
WHERE name LIKE 'orbeon_f_a_a_1%';
```

We can get the list of the views and their columns with the following SQL.

```sql
SELECT o.name AS view_name,
       c.name AS column_name,
       t.name AS data_type,
       c.max_length
FROM sys.columns c
JOIN sys.objects o ON c.object_id = o.object_id
JOIN sys.types t ON c.user_type_id = t.user_type_id
WHERE o.type = 'V'
  AND o.name LIKE 'orbeon_f_a_a_1%'
ORDER BY o.name, c.column_id;
```

On 2024.1 with `fully-qualified-names` set to `false`:

```
     view_name     |         column_name         | data_type | max_length 
-------------------+-----------------------------+-----------+------------
 orbeon_f_a_a_1    | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1    | metadata_created            | datetime2 |          8 
 orbeon_f_a_a_1    | metadata_last_modified_time | datetime2 |          8 
 orbeon_f_a_a_1    | metadata_last_modified_by   | nvarchar  |        510 
 orbeon_f_a_a_1    | s1g1f1                      | varchar   |       8000 
 orbeon_f_a_a_1_g2 | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1_g2 | g2_repetition               | bigint    |          8 
 orbeon_f_a_a_1_g2 | s1g2f2                      | varchar   |       8000 
 orbeon_f_a_a_1_s2 | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1_s2 | s2_repetition               | bigint    |          8 
 orbeon_f_a_a_1_s2 | f3                          | varchar   |       8000 
```

On 2024.1 with `fully-qualified-names` set to `true`:

```
           view_name            |         column_name         | data_type | max_length 
--------------------------------+-----------------------------+-----------+------------
 orbeon_f_a_a_1                 | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1                 | metadata_created            | datetime2 |          8 
 orbeon_f_a_a_1                 | metadata_last_modified_time | datetime2 |          8 
 orbeon_f_a_a_1                 | metadata_last_modified_by   | nvarchar  |        510 
 orbeon_f_a_a_1                 | s1_s1g1f1                   | varchar   |       8000 
 orbeon_f_a_a_1_s1_g2           | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1_s1_g2           | g2_repetition               | bigint    |          8 
 orbeon_f_a_a_1_s1_g2           | s1g2f2                      | varchar   |       8000 
 orbeon_f_a_a_1_s2_s2_iteration | metadata_document_id        | nvarchar  |        510 
 orbeon_f_a_a_1_s2_s2_iteration | s2_repetition               | bigint    |          8 
 orbeon_f_a_a_1_s2_s2_iteration | f3                          | varchar   |       8000 
```

On 2023.1.5:

```
   view_name    |         column_name         | data_type | max_length 
----------------+-----------------------------+-----------+------------
 ORBEON_F_A_A_1 | METADATA_DOCUMENT_ID        | nvarchar  |        510 
 ORBEON_F_A_A_1 | METADATA_CREATED            | datetime2 |          8 
 ORBEON_F_A_A_1 | METADATA_LAST_MODIFIED_TIME | datetime2 |          8 
 ORBEON_F_A_A_1 | METADATA_LAST_MODIFIED_BY   | nvarchar  |        510 
 ORBEON_F_A_A_1 | S1_S1G1F1                   | nvarchar  |        510 
```

**Comment by avernet** [comment]
- **Suggested change**: when `fully-qualified-names` is set to `true`, also prefix the column names of the sub tables with the section name.
- **Impact in the above examples**: in the B case
  - `s/s1g2_repetition/s1_s1g2_repetition/`
  - `s/s1g2f2/s1_s1g2f2/`
  - `s/s2g3f3/s2_s2g3f3/`

**Comment by avernet** [comment]
![image](https://github.com/user-attachments/assets/41e05aa0-692e-442e-9a51-c6c11c635892)

**Comment by obruchez** [comment]
- [ ] rename property
- [ ] prefix all columns
- [ ] do not include iteration name

**Comment by obruchez** [comment]
A few months later, I don't know why exactly I insisted on including the iteration name in the view/column names. Maybe this came up in a discussion?

The code seems to be wrong anyway, as the iteration name is built from `DefaultIterationSuffix` instead of being extracted from the definition/binding.

The following code:

```scala
val addIterationNameToPath = section && repeated
val iterationNameOpt       = addIterationNameToPath.flatOption(nameToAddToPathOpt.map(_ + DefaultIterationSuffix))
val currentPath            = path ++ nameToAddToPathOpt.toList ++ iterationNameOpt.toList
```

can simply become:

```scala
val currentPath            = path ++ nameToAddToPathOpt.toList
```

**Comment by obruchez** [comment]
I'm a bit less certain about the other point (i.e. prefixing all view/column names with the full name/path if `fully-qualified-name=true`). On the one hand, consistency is good. On the other hand, we will introduce unnecessary long names that were not present before (since we didn't have flat views for repeated grids/sections). And we already have this rather confusing shortening algorithm that produce names that are hard to parse by human. In your example, we have short section/grid/control names, but in real cases, the names will be longer and will probably have to be shortened.

If we decide to prefix everything if `fully-qualified-name=true`, we'll have to conditionally call `FlatView.relativePath` (two locations, one for view names, one for column names). We'll also have to prefix the repetition number column.

**Comment by obruchez** [comment]
Also, regarding the property name. The idea is that we include the whole "path"/prefix in the names if `fully-qualified-name=true`. This works for nested sections/grids. I know that in real case scenarios, we probably don't have many situations with nested sections/grids (i.e. more than 2 levels). but `prefix-with-section-name` seems to imply that only the closed section will be included in the name, whereas all "ancestor" sections are included. Repeated grid names are also included.

**Comment by obruchez** [comment]
Changed the property name from `oxf.fr.persistence.[provider].flat-view.fully-qualified-names` to `oxf.fr.persistence.[provider].flat-view.prefixes-in-main-view-column-names`. This property doesn't change the behavior of sub-views. When `true`, it makes main views behave like in 2023.x.

**Comment by obruchez** [comment]
Example.

2023 behavior:

<img width="347" alt="2023 x" src="https://github.com/user-attachments/assets/f97bc86d-5862-4723-a3c0-682ceea825d7">

2024 _before_ latest fix, with property = `false`:

<img width="341" alt="2024 1-before-false" src="https://github.com/user-attachments/assets/653703ce-6709-42a7-8697-3734109e2d2d">

2024 _before_ latest fix, with property = `true`:

<img width="335" alt="2024 1-before-true" src="https://github.com/user-attachments/assets/023641a2-0689-452c-9fa8-2508db5668c0">

2024 _after_ latest fix, with property = `false`:

<img width="337" alt="2024 1-after-false" src="https://github.com/user-attachments/assets/ae7b4238-2466-436c-997a-3d749333f893">

2024 _after_ latest fix, with property = `true`:

<img width="339" alt="2024 1-after-true" src="https://github.com/user-attachments/assets/c15e4b67-be22-45ee-bfa7-8616cc30a6ba">
